crates/miden-lib/src/transaction/inputs.rs

```diff
@@ -178,251 +27,408 @@ enum AccountInputsType {
+                            e,
                         )
-                        .map_err(|err| {
-                            TransactionInputError::InvalidMerklePath(
-                                format!("input note ID {}", note.id()).into(),
-                                err,
-                            )
-                        })?,
-                );
-                note_data.push(proof.location().block_num().into());
-                note_data.extend(note_block_header.sub_commitment());
-                note_data.extend(note_block_header.note_root());
-                note_data.push(proof.location().node_index_in_block().into());
+                    })?,
+            );
+            // populate advice map with Sparse Merkle Tree leaf nodes
+            self.extend_map(core::iter::once((proof.leaf().hash(), proof.leaf().to_elements())));
+        }
+
+        // --- account code -------------------------------------------------------
+
+        let code = account.code();
+        self.extend_map([
+            // CODE_COMMITMENT -> [[ACCOUNT_PROCEDURE_DATA]]
+            (code.commitment(), code.as_elements()),
+        ]);
+
+        // --- account vault ------------------------------------------------------
+
+        self.extend_merkle_store(account.vault().inner_nodes());
+        // populate advice map with Sparse Merkle Tree leaf nodes
+        self.extend_map(account.vault().leaves().map(|leaf| (leaf.hash(), leaf.to_elements())));
+
+        // --- account state ------------------------------------------------------
+
+        let acc_id = account.id();
+        let account_id_key: Digest =
+            Digest::from([acc_id.suffix(), acc_id.prefix().as_felt(), ZERO, ZERO]);
+        match input_type {
+            // if a seed was provided, extend the map appropriately
+            AccountInputsType::Native(Some(seed)) => {
+                self.extend_map([
+                    // ACCOUNT_ID -> ACCOUNT_SEED
+                    (account_id_key, seed.to_vec()),
+                ]);
             },
-            InputNote::Unauthenticated { .. } => {
-                // NOTE: keep in sync with the `prologue::process_input_note` kernel procedure
-                // Push the `is_authenticated` flag
-                note_data.push(Felt::ZERO);
+            // for foreign accounts, we need to insert the id to state mapping
+            AccountInputsType::Foreign => {
+                // NOTE: keep this in sync with the start_foreign_context kernel procedure
+                let header = AccountHeader::from(account);
+                self.extend_map([
+                    // ACCOUNT_ID -> [ID_AND_NONCE, VAULT_ROOT, STORAGE_COMMITMENT,
+                    // CODE_COMMITMENT]
+                    (account_id_key, header.as_elements()),
+                ]);
             },
+            // native non-new accounts do not need extra inputs
+            AccountInputsType::Native(None) => {},
```

Comment:
I would maybe move this logic up into the constructor. This should also eliminate the need for the `AccountInputsType` enum.

Reply:
Moved this. Not sure the result is great because now it's responsibility of the caller to not to forget to add the extra inputs, but at least the enum is gone which is nice.